eventrate/eventrate.go

```diff
@@ -0,0 +1,111 @@
+package eventrate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	eventspb "code.vegaprotocol.io/protos/vega/events/v1"
+	"code.vegaprotocol.io/vegatools/stream"
+	"google.golang.org/protobuf/proto"
+)
+
+// Opts are the command line options passed to the sub command
+type Opts struct {
+	ServerAddr       string
+	Buckets          int
+	SecondsPerBucket int
+}
+
+func min(a, b uint64) uint64 {
+	if a > b {
+		return b
+	}
+	return a
+}
+
+func max(a, b uint64) uint64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+type data struct {
+	Events uint64
+	Bytes  uint64
+}
+
+func fixUnits(bytes uint64) string {
+	if bytes > 1024*1024 {
+		return fmt.Sprintf("%dMB", bytes/(1024*1024))
+	} else if bytes > 1024 {
+		return fmt.Sprintf("%dKB", bytes/1024)
+	}
+	return fmt.Sprintf("%dB", bytes)
+}
+
+// Run is the main function of `eventrate` package
+func Run(opts Opts) error {
+	var dataThisSecond data
+	var historicData []data
+	var mu sync.Mutex
+
+	if len(opts.ServerAddr) <= 0 {
+		return fmt.Errorf("error: missing grpc server address")
+	}
+
+	handleEvent := func(e *eventspb.BusEvent) {
+		mu.Lock()
+		dataThisSecond.Events++
+		dataThisSecond.Bytes += uint64(proto.Size(e))
+		mu.Unlock()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg := sync.WaitGroup{}
+	if err := stream.ReadEvents(ctx, cancel, &wg, 0, "", "", opts.ServerAddr, handleEvent, true, nil); err != nil {
+		return fmt.Errorf("error reading events: %v", err)
+	}
+
+	for {
+		time.Sleep(time.Second * time.Duration(opts.SecondsPerBucket))
+		mu.Lock()
+		historicData = append(historicData, dataThisSecond)
+		dataThisSecond.Events = 0
+		dataThisSecond.Bytes = 0
+		mu.Unlock()
+
+		// Cap the number of historic buckets we keep
+		if len(historicData) > opts.Buckets {
+			historicData = historicData[1:]
+		}
+
+		// Calculate values
+		minEvents := historicData[0].Events
+		maxEvents := historicData[0].Events
+		minBytes := historicData[0].Bytes
+		maxBytes := historicData[0].Bytes
```

Comment:
we could probably just declare these variables instead of instantiating them, because when ranging over `historicData` below, we also include the first element

Reply:
If we don't set the values outside the loop then the default value for each variable will be used on the first loop and the min calculations will always return 0.